core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/FastRdfsForwardChainingSail.java

```diff
@@ -0,0 +1,311 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.StampedLock;
+import java.util.stream.Collectors;
+
+/**
+ * @author Håvard Mikkelsen Ottestad
+ */
+
+public class FastRdfsForwardChainingSail extends AbstractForwardChainingInferencer {
+
+    NotifyingSail data;
+    Repository schema;
+
+    static final Random random = new Random();
+
+    private StampedLock readWriteLock = new StampedLock();
+
+    boolean useAllRdfsRules = true;
+
+    List<Statement> subClassOfStatements = new ArrayList<>();
+    Set<Resource> properties = new HashSet<>();
+    List<Statement> subPropertyOfStatements = new ArrayList<>();
+    List<Statement> rangeStatements = new ArrayList<>();
+    List<Statement> domainStatements = new ArrayList<>();
+
+
+    Map<Resource, Set<Resource>> calculatedTypes = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedProperties = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedRange = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedDomain = new HashMap<>();
+    private boolean sharedSchema;
+
+    void clearInferenceTables() {
+        subClassOfStatements = new ArrayList<>();
+        properties = new HashSet<>();
+        subPropertyOfStatements = new ArrayList<>();
+        rangeStatements = new ArrayList<>();
+        domainStatements = new ArrayList<>();
+        calculatedTypes = new HashMap<>();
+        calculatedProperties = new HashMap<>();
+        calculatedRange = new HashMap<>();
+        calculatedDomain = new HashMap<>();
+    }
+
+    void readLock(FastRdfsForwardChainingSailConnection connection) {
+
+        if (numberOfThreadsWaitingForWriteLock.get() > 0) {
+//            System.err.println("starve reads");
+        }
+
+        while (numberOfThreadsWaitingForWriteLock.get() > 0) {
+            Thread.yield();
+        }
+
+        if (connection.lockStamp != 0) {
+            throw new IllegalStateException("Connection already has a lock!");
+        }
+        connection.lockStamp = readWriteLock.readLock();
+//        System.err.println("readLock: " + connection.lockStamp);
+
+    }
+
+    void releaseLock(FastRdfsForwardChainingSailConnection connection) {
+        readWriteLock.unlock(connection.lockStamp);
+//        System.err.println("Released lock: " + connection.lockStamp);
+
+        connection.lockStamp = 0;
+    }
+
+
+    AtomicInteger numberOfThreadsWaitingForWriteLock = new AtomicInteger(0);
+
+    void upgradeLock(FastRdfsForwardChainingSailConnection connection) {
+
+////        System.err.println("Attempt writelock: "+connection.lockStamp);
+
+        numberOfThreadsWaitingForWriteLock.incrementAndGet();
+
+        try {
+           while(true) {
+                long l = readWriteLock.tryConvertToWriteLock(connection.lockStamp);
```

Comment:
Why are you looping and sleeping, rather than just doing a blocking call on `readWriteLock.writeLock()`?